codalab/worker_manager/worker_manager.py

```diff
@@ -0,0 +1,167 @@
+import logging
+import time
+from codalab.lib.codalab_manager import CodaLabManager
+from codalab.worker.bundle_state import State
+from collections import namedtuple
+
+logger = logging.getLogger(__name__)
+
+# Represents a AWS/Azure job that runs a single cl-worker.
+# For now, we don't need to keep any information.
+WorkerJob = namedtuple('WorkerJob', [])
+
+
+class WorkerManager(object):
+    """
+    The abstract class for a worker manager.  Different backends like AWS,
+    Azure, Slurm should override `get_worker_jobs` and `start_worker_job`.
+
+    The basic architecture of the WorkerManager is extremely simple: to a
+    first-order approximation, it simply launches `cl-worker`s as AWS/Azure
+    batch jobs as long as there staged bundles.
+
+    The simplicity means that we don't need to manage state about how workers
+    and bundles are related - that logic is complex and is done by the usual
+    worker system, and we don't want to duplicate that for every possible
+    worker backend.
+
+    More specifically, a worker manager will monitor a job queue to see how
+    many worker jobs are running, and try to keep that between `min_workers`
+    and `max_workers`.  It will also monitor the staged bundles that satisfy a
+    certain `search` criterion.  If there are staged bundles then it will issue
+    a `start_worker_job()` call, provided some other conditions are met (e.g.,
+    don't start workers too fast).
+
+    The WorkerManager is all client-side code, so it can be customized as one
+    sees fit.
+
+    Notes:
+    - The worker manager is not visible via CodaLab.  One needs to monitor the
```

Comment:
What specifically do you mean by "not visible via CodaLab"? One can still view the docker containers when running the system.

Reply:
What I mean is that you can't see from CodaLab that the worker manager is trying to do something (e.g., trying to launch jobs) or even whether it died.